File: apps/macos/Sources/AgentCafeMac/SidecarClient.swift
import Foundation

struct SidecarRunResult {
    let report: DiagnosticReport?
    let errorCode: String?
    let errorMessage: String?

    static func success(_ report: DiagnosticReport) -> SidecarRunResult {
        SidecarRunResult(report: report, errorCode: nil, errorMessage: nil)
    }

    static func failure(_ code: String, _ message: String) -> SidecarRunResult {
        SidecarRunResult(report: nil, errorCode: code, errorMessage: message)
    }
}

final class SidecarClient {
    private let sidecarPath: String
    private let decoder: JSONDecoder
    private var nextId = 0

    init(sidecarPath: String? = nil) {
        self.sidecarPath = sidecarPath
            ?? ProcessInfo.processInfo.environment["AGENTCAFE_SIDECAR"]
            ?? SidecarClient.defaultSidecarPath()
        self.decoder = JSONDecoder()
        self.decoder.dateDecodingStrategy = .iso8601
    }

    func runDoctor() async -> SidecarRunResult {
        if let fixture = ProcessInfo.processInfo.environment["AGENTCAFE_UI_FIXTURE"], !fixture.isEmpty {
            return await loadFixture(path: fixture)
        }

        guard FileManager.default.fileExists(atPath: sidecarPath) else {
            return .failure("sidecar_missing", "Sidecar not found at \(sidecarPath).")
        }

        let process = Process()
        process.executableURL = URL(fileURLWithPath: sidecarPath)

        let input = Pipe()
        let output = Pipe()
        let error = Pipe()
        process.standardInput = input
        process.standardOutput = output
        process.standardError = error

        do {
            try process.run()
            try writeRequest(
                method: "ipc.handshake",
                params: [
                    "protocol_version": "1.0",
                    "ui_name": "agentcafe-macos-swiftui",
                    "ui_version": "0.1.0",
                    "ui_platform": "macos",
                    "ui_capabilities": [],
                    "nonce": UUID().uuidString
                ],
                pipe: input
            )

            let handshake = try readEnvelope(from: output)
            if let error = handshake.error {
                terminate(process)
                return .failure(error.data?.code ?? "handshake_failed", error.message)
            }

            try writeRequest(method: "doctor.run", params: [:], pipe: input)
            let doctor = try readEnvelope(from: output)
            input.fileHandleForWriting.closeFile()

            if let error = doctor.error {
                terminate(process)
                return .failure(error.data?.code ?? "doctor_failed", error.message)
            }

            guard let result = doctor.result else {
                terminate(process)
                return .failure("doctor_failed", "doctor.run returned no result.")
            }

            let report = try decoder.decode(DiagnosticReport.self, from: result)
            return .success(report)
        } catch {
            terminate(process)
            return .failure("sidecar_crash", error.localizedDescription)
        }
    }

    private func loadFixture(path: String) async -> SidecarRunResult {
        do {
            let data = try Data(contentsOf: URL(fileURLWithPath: path))
            let report = try decoder.decode(DiagnosticReport.self, from: data)
            return .success(report)
        } catch {
            return .failure("fixture_invalid", error.localizedDescription)
        }
    }

    private func writeRequest(method: String, params: [String: Any], pipe: Pipe) throws {
        nextId += 1
        let request: [String: Any] = [
            "jsonrpc": "2.0",
            "id": String(nextId),
            "method": method,
            "params": params
        ]
        let data = try JSONSerialization.data(withJSONObject: request)
        pipe.fileHandleForWriting.write(data)
        pipe.fileHandleForWriting.write(Data("\n".utf8))
    }

    private func readEnvelope(from pipe: Pipe) throws -> RpcEnvelope {
        guard let line = pipe.fileHandleForReading.readLine() else {
            throw SidecarClientError.closedStream
        }
        return try JSONDecoder().decode(RpcEnvelope.self, from: line)
    }

    private func terminate(_ process: Process) {
        if process.isRunning {
            process.terminate()
        }
    }

    private static func defaultSidecarPath() -> String {
        let root = findRepositoryRoot(start: FileManager.default.currentDirectoryPath)
        return URL(fileURLWithPath: root)
            .appendingPathComponent("target")
            .appendingPathComponent("debug")
            .appendingPathComponent("agentcafe-sidecar")
            .path
    }

    private static func findRepositoryRoot(start: String) -> String {
        var url = URL(fileURLWithPath: start)
        while url.path != "/" {
            let cargo = url.appendingPathComponent("Cargo.toml").path
            let core = url.appendingPathComponent("core").path
            if FileManager.default.fileExists(atPath: cargo)
                && FileManager.default.fileExists(atPath: core) {
                return url.path
            }
            url.deleteLastPathComponent()
        }
        return start
    }
}

private struct RpcEnvelope: Decodable {
    let result: Data?
    let error: RpcError?

    enum CodingKeys: String, CodingKey {
        case result
        case error
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        if container.contains(.result) {
            let object = try container.decode(AnyDecodable.self, forKey: .result)
            result = try JSONSerialization.data(withJSONObject: object.value)
        } else {
            result = nil
        }
        error = try container.decodeIfPresent(RpcError.self, forKey: .error)
    }
}

private struct RpcError: Decodable {
    let code: Int
    let message: String
    let data: RpcErrorData?
}

private struct RpcErrorData: Decodable {
    let code: String?
    let stage: String?
}

private struct AnyDecodable: Decodable {
    let value: Any

    init(from decoder: Decoder) throws {
        if let container = try? decoder.singleValueContainer() {
            if container.decodeNil() {
                value = NSNull()
            } else if let bool = try? container.decode(Bool.self) {
                value = bool
            } else if let int = try? container.decode(Int.self) {
                value = int
            } else if let double = try? container.decode(Double.self) {
                value = double
            } else if let string = try? container.decode(String.self) {
                value = string
            } else if let array = try? container.decode([AnyDecodable].self) {
                value = array.map(\.value)
            } else if let dictionary = try? container.decode([String: AnyDecodable].self) {
                value = dictionary.mapValues(\.value)
            } else {
                throw DecodingError.dataCorruptedError(
                    in: container,
                    debugDescription: "Unsupported JSON value."
                )
            }
        } else {
            throw SidecarClientError.invalidEnvelope
        }
    }
}

private enum SidecarClientError: Error {
    case closedStream
    case invalidEnvelope
}

private extension FileHandle {
    func readLine() -> Data? {
        var data = Data()
        while true {
            let byte = self.readData(ofLength: 1)
            if byte.isEmpty {
                return data.isEmpty ? nil : data
            }
            if byte == Data("\n".utf8) {
                return data
            }
            data.append(byte)
        }
    }
}

File: apps/windows-wpf/Services/SidecarClient.cs
using System.Diagnostics;
using System.IO;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.Json.Serialization;
using AgentCafe.Windows.Models;

namespace AgentCafe.Windows.Services;

public sealed class SidecarClient
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull
    };
    private static readonly JsonSerializerOptions PrettyJsonOptions = new()
    {
        WriteIndented = true
    };

    private readonly string _sidecarPath;
    private int _nextId;

    public SidecarClient(string? sidecarPath = null)
    {
        _sidecarPath = sidecarPath
            ?? Environment.GetEnvironmentVariable("AGENTCAFE_SIDECAR")
            ?? ResolveDefaultSidecarPath();
    }

    public async Task<SidecarRunResult<DiagnosticReport>> RunDoctorAsync(
        CancellationToken cancellationToken)
    {
        if (Environment.GetEnvironmentVariable("AGENTCAFE_UI_FIXTURE") is { Length: > 0 } fixture)
        {
            return await LoadFixtureAsync(fixture, cancellationToken);
        }

        if (!File.Exists(_sidecarPath))
        {
            return SidecarRunResult<DiagnosticReport>.Failure(
                "sidecar_missing",
                $"Sidecar not found at {_sidecarPath}.");
        }

        using var process = StartSidecar();
        try
        {
            await SendAsync(process, "ipc.handshake", new
            {
                protocol_version = "1.0",
                ui_name = "agentcafe-windows-wpf",
                ui_version = "0.1.0",
                ui_platform = "windows",
                ui_capabilities = Array.Empty<string>(),
                nonce = Guid.NewGuid().ToString("N")
            }, cancellationToken);

            var handshake = await ReadResponseAsync<JsonElement>(process, cancellationToken);
            if (!handshake.IsSuccess)
            {
                return SidecarRunResult<DiagnosticReport>.Failure(
                    handshake.ErrorCode ?? "handshake_failed",
                    handshake.ErrorMessage ?? "Handshake failed.");
            }

            await SendAsync(process, "doctor.run", new { }, cancellationToken);
            var doctor = await ReadDoctorResponseAsync(process, cancellationToken);
            if (!doctor.IsSuccess || doctor.Result is null)
            {
                return SidecarRunResult<DiagnosticReport>.Failure(
                    doctor.ErrorCode ?? "doctor_failed",
                    doctor.ErrorMessage ?? "doctor.run failed.");
            }

            return SidecarRunResult<DiagnosticReport>.Success(doctor.Result, doctor.PrettyJson ?? "");
        }
        catch (OperationCanceledException)
        {
            TryKill(process);
            return SidecarRunResult<DiagnosticReport>.Failure(
                "timeout",
                "Sidecar request timed out.");
        }
        catch (Exception ex) when (ex is IOException or InvalidOperationException)
        {
            return SidecarRunResult<DiagnosticReport>.Failure("sidecar_crash", ex.Message);
        }
        finally
        {
            if (!process.HasExited)
            {
                process.StandardInput.Close();
            }
        }
    }

    private Process StartSidecar()
    {
        var process = new Process
        {
            StartInfo = new ProcessStartInfo
            {
                FileName = _sidecarPath,
                RedirectStandardInput = true,
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                UseShellExecute = false,
                CreateNoWindow = true
            },
            EnableRaisingEvents = true
        };
        if (!process.Start())
        {
            throw new InvalidOperationException("Unable to start sidecar.");
        }
        return process;
    }

    private async Task SendAsync(
        Process process,
        string method,
        object parameters,
        CancellationToken cancellationToken)
    {
        var id = Interlocked.Increment(ref _nextId).ToString();
        var request = new
        {
            jsonrpc = "2.0",
            id,
            method,
            @params = parameters
        };
        var line = JsonSerializer.Serialize(request, JsonOptions);
        await process.StandardInput.WriteLineAsync(line.AsMemory(), cancellationToken);
        await process.StandardInput.FlushAsync(cancellationToken);
    }

    private static async Task<RpcReadResult<T>> ReadResponseAsync<T>(
        Process process,
        CancellationToken cancellationToken)
    {
        var line = await process.StandardOutput.ReadLineAsync(cancellationToken);
        if (string.IsNullOrWhiteSpace(line))
        {
            return RpcReadResult<T>.Failure("sidecar_crash", "Sidecar closed stdout.");
        }

        var envelope = JsonSerializer.Deserialize<RpcEnvelope<T>>(line, JsonOptions);
        if (envelope?.Error is not null)
        {
            return RpcReadResult<T>.Failure(
                envelope.Error.Data?.Code ?? "sidecar_error",
                envelope.Error.Message);
        }
        return RpcReadResult<T>.Success(envelope!.Result);
    }

    private static async Task<RpcReadResult<DiagnosticReport>> ReadDoctorResponseAsync(
        Process process,
        CancellationToken cancellationToken)
    {
        var line = await process.StandardOutput.ReadLineAsync(cancellationToken);
        if (string.IsNullOrWhiteSpace(line))
        {
            return RpcReadResult<DiagnosticReport>.Failure("sidecar_crash", "Sidecar closed stdout.");
        }

        var envelope = JsonSerializer.Deserialize<RpcEnvelope<JsonElement>>(line, JsonOptions);
        if (envelope?.Error is not null)
        {
            return RpcReadResult<DiagnosticReport>.Failure(
                envelope.Error.Data?.Code ?? "sidecar_error",
                envelope.Error.Message);
        }

        var resultJson = envelope!.Result.GetRawText();
        var report = JsonSerializer.Deserialize<DiagnosticReport>(resultJson, JsonOptions);
        var prettyJson = JsonNode.Parse(resultJson)?.ToJsonString(PrettyJsonOptions) ?? resultJson;
        return RpcReadResult<DiagnosticReport>.Success(report, prettyJson);
    }

    private static async Task<SidecarRunResult<DiagnosticReport>> LoadFixtureAsync(
        string fixture,
        CancellationToken cancellationToken)
    {
        var json = await File.ReadAllTextAsync(fixture, cancellationToken);
        var report = JsonSerializer.Deserialize<DiagnosticReport>(json, JsonOptions);
        var prettyJson = JsonNode.Parse(json)?.ToJsonString(PrettyJsonOptions) ?? json;
        return report is null
            ? SidecarRunResult<DiagnosticReport>.Failure("fixture_invalid", "Fixture did not decode.")
            : SidecarRunResult<DiagnosticReport>.Success(report, prettyJson);
    }

    private static string ResolveDefaultSidecarPath()
    {
        var root = FindRepositoryRoot(AppContext.BaseDirectory);
        var fileName = OperatingSystem.IsWindows()
            ? "agentcafe-sidecar.exe"
            : "agentcafe-sidecar";
        return Path.Combine(root, "target", "debug", fileName);
    }

    private static string FindRepositoryRoot(string start)
    {
        var directory = new DirectoryInfo(start);
        while (directory is not null)
        {
            if (File.Exists(Path.Combine(directory.FullName, "Cargo.toml"))
                && Directory.Exists(Path.Combine(directory.FullName, "core")))
            {
                return directory.FullName;
            }
            directory = directory.Parent;
        }

        return Path.GetFullPath(Path.Combine(AppContext.BaseDirectory, "..", "..", "..", ".."));
    }

    private static void TryKill(Process process)
    {
        try
        {
            if (!process.HasExited)
            {
                process.Kill(entireProcessTree: true);
            }
        }
        catch
        {
            // Best-effort cleanup after timeout or crash.
        }
    }

    private sealed record RpcEnvelope<T>(
        [property: JsonPropertyName("result")] T? Result,
        [property: JsonPropertyName("error")] RpcError? Error
    );

    private sealed record RpcError(
        [property: JsonPropertyName("code")] int Code,
        [property: JsonPropertyName("message")] string Message,
        [property: JsonPropertyName("data")] RpcErrorData? Data
    );

    private sealed record RpcErrorData(
        [property: JsonPropertyName("code")] string? Code,
        [property: JsonPropertyName("stage")] string? Stage
    );

    private sealed record RpcReadResult<T>(
        bool IsSuccess,
        T? Result,
        string? PrettyJson,
        string? ErrorCode,
        string? ErrorMessage)
    {
        public static RpcReadResult<T> Success(T? result, string? prettyJson = null) =>
            new(true, result, prettyJson, null, null);
        public static RpcReadResult<T> Failure(string errorCode, string errorMessage) =>
            new(false, default, null, errorCode, errorMessage);
    }
}

public sealed record SidecarRunResult<T>(
    bool IsSuccess,
    T? Report,
    string? PrettyJson,
    string? ErrorCode,
    string? ErrorMessage)
{
    public static SidecarRunResult<T> Success(T report, string prettyJson) =>
        new(true, report, prettyJson, null, null);
    public static SidecarRunResult<T> Failure(string errorCode, string errorMessage) =>
        new(false, default, null, errorCode, errorMessage);
}
